pkg/node/manager/events_test.go

```diff
@@ -0,0 +1,101 @@
+//go:build unit || !integration
+
+package manager_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	gomock "go.uber.org/mock/gomock"
+
+	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"github.com/bacalhau-project/bacalhau/pkg/node/manager"
+	"github.com/stretchr/testify/suite"
+)
+
+type EventEmitterSuite struct {
+	suite.Suite
+	ctrl  *gomock.Controller
+	ctx   context.Context
+	clock *clock.Mock
+}
+
+func TestEventEmitterSuite(t *testing.T) {
+	suite.Run(t, new(EventEmitterSuite))
+}
+
+func (s *EventEmitterSuite) SetupTest() {
+	s.ctrl = gomock.NewController(s.T())
+	s.ctx = context.Background()
+	s.clock = clock.NewMock()
+}
+
+func (s *EventEmitterSuite) TestNewNodeEventEmitter() {
+	e := manager.NewNodeEventEmitter()
+	s.NotNil(e)
+
+	mockHandler := manager.NewMockNodeEventHandler(s.ctrl)
+	mockHandler.EXPECT().HandleNodeEvent(gomock.Any(), gomock.Any(), manager.NodeEventApproved)
+
+	e.RegisterHandler(mockHandler)
+
+	err := e.EmitEvent(s.ctx, models.NodeInfo{}, manager.NodeEventApproved)
+	s.NoError(err)
+}
+
+func (s *EventEmitterSuite) TestRegisterCallback() {
+	e := manager.NewNodeEventEmitter()
+	s.NotNil(e)
+
+	mockHandler := manager.NewMockNodeEventHandler(s.ctrl)
+	e.RegisterHandler(mockHandler)
+}
+
+func (s *EventEmitterSuite) TestEmitEvent() {
+	e := manager.NewNodeEventEmitter()
+	s.NotNil(e)
+
+	mockHandler := manager.NewMockNodeEventHandler(s.ctrl)
+	mockHandler.EXPECT().HandleNodeEvent(gomock.Any(), gomock.Any(), manager.NodeEventApproved)
+	mockHandler.EXPECT().HandleNodeEvent(gomock.Any(), gomock.Any(), manager.NodeEventRejected)
+
+	e.RegisterHandler(mockHandler)
+
+	err := e.EmitEvent(s.ctx, models.NodeInfo{}, manager.NodeEventApproved)
+	s.NoError(err)
+
+	err = e.EmitEvent(s.ctx, models.NodeInfo{}, manager.NodeEventRejected)
+	s.NoError(err)
+}
+
+func (s *EventEmitterSuite) TestEmitEventWithNoCallbacks() {
+	e := manager.NewNodeEventEmitter()
+	s.NotNil(e)
+
+	err := e.EmitEvent(s.ctx, models.NodeInfo{}, manager.NodeEventApproved)
+	s.NoError(err)
+}
+
+func (s *EventEmitterSuite) TestEmitWithSlowCallback() {
+	e := manager.NewNodeEventEmitter(manager.WithClock(s.clock))
+	s.NotNil(e)
+
+	e.RegisterHandler(testSleepyHandler{s.clock})
+
+	go func() {
+		s.clock.Add(10 * time.Second)
+	}()
```

Comment:
why does this need a go routine? I'd assume adding 10 seconds to the clock is non-blocking?

Reply:
Some weirdness with the mock clock where it won't work when it's inline.  I've been trying to work out if I'm just holding it wrong but I suspect it's something to do with the work in EmitEvent happening in a goroutine and requiring a yield (but I'm not 100% sure)